lib/Source/AutoHog/AutoHogImporter.cpp

```diff
@@ -0,0 +1,516 @@
+          ioPort.connectedInputPort = connectedPortName;
+        }
+      }
+    }
+
+    ports[ioPort.portName] = ioPort;
+  }
+}
+
+/// Parse the connections field of a cell object in the JSON document.
+/// - Populates the connections map, mapping the local port name to the
+/// connected
+///   Port object (an IO port or another cell).
+/// - Populates the outputPortToConnectedCellPort map, mapping an output ioPort
+///   to the corresponding local CELL type Port of the current cell.
+void parseCellConnections(const char *cellName,
+                          const rapidjson::Value &cellData,
+                          const StringMap<Port> &ioPorts,
+                          StringMap<Port> &connections,
+                          StringMap<Port> &outputPortToConnectedCellPort) {
+  LLVM_DEBUG(llvm::dbgs() << "Parsing cell connections\n");
+  auto &jsonConnections = cellData["connections"];
+  assert(jsonConnections.IsObject() &&
+         "Expected 'connections' to be an object");
+  for (rapidjson::Value::ConstMemberIterator itr =
+           jsonConnections.MemberBegin();
+       itr != jsonConnections.MemberEnd(); ++itr) {
+    const char *localPortName = itr->name.GetString();
+    const rapidjson::Value &portData = itr->value;
+    assert(portData.IsObject() &&
+           "Expected 'connections[...]' to be an object");
+    const char *connectedPortName = portData["port"].GetString();
+
+    Port port;
+    port.portName = connectedPortName;
+
+    if (portData.HasMember("cell")) {
+      port.type = PortType::CELL;
+      const char *connectedCell = portData["cell"].GetString();
+      port.cellName = connectedCell;
+    } else {
+      port.type = ioPorts.at(connectedPortName).type;
+      if (port.type == PortType::OUTPUT) {
+        Port localPort(PortType::CELL, std::string(localPortName),
+                       std::string(cellName));
+        outputPortToConnectedCellPort[StringRef(connectedPortName)] = localPort;
+      }
+    }
+    connections[StringRef(localPortName)] = port;
+  }
+}
+
+/// Construct the cell graph from the JSON document.
+/// - Returns a graph::Graph<const char *> object where the nodes are cell names
+///   and the edges are dependencies between cells.
+graph::Graph<std::string> constructCellGraph(
+    const rapidjson::Document &document) {
+  LLVM_DEBUG(llvm::dbgs() << "Topologically sorting cells\n");
+  graph::Graph<std::string> cellGraph;
+  auto &cells = document["cells"];
+  assert(cells.IsObject() && "Expected 'cells' to be an object");
+  for (rapidjson::Value::ConstMemberIterator itr = cells.MemberBegin();
+       itr != cells.MemberEnd(); ++itr) {
+    const rapidjson::Value &cell = itr->value;
+    assert(cell.IsObject() && "Expected cell to be an object");
+    const char *cellName = cell["cell_name"].GetString();
+    cellGraph.addVertex(std::string(cellName));
+    auto &connections = cell["connections"];
+    assert(connections.IsObject() && "Expected 'connections' to be an object");
```

Comment:
same with the comment here and below - these assertions seem like they're catching JSON inputs that would be invalid, and that these would be skipped in release mode and probably cause unexpected confusion. it *might* be worth putting some kind of input error here instead? or we can i suppose always hope the input will be valid..

Reply:
I believe most of these type assertions would fail on the very next line when iteration is attempted. I'm OK leaving them as debug assertions for now.